[feature] Remove Blank Profile Objects

billyhunt opened the pull request:
Fixes: #2412 and #2384

A resubmission of  #3515

Purpose
-----------

Originally, this PR was a fix for an Internal Server Error (500) when Clicking "Submit" in the Profile when there is a blank School or Job object.  This problem has already been fixed in another PR.  The same error was also occurring in the required "Full Name" field in the "Name" section of the profile, and is addressed in this PR.

<img width="747" alt="screen shot 2015-08-25 at 9 33 03 am" src="https://cloud.githubusercontent.com/assets/6232068/9467838/5546d578-4b0c-11e5-913e-6a2629ae0e2b.png">

The PR was also expanded to improve usability of the Education and Employment tabs, specifically to allow users to remove all jobs and schools from their profile.  (not currently possible because a user cannot save if the Institution field is not filled out)

Changes
------------

2.  The "Full Name" required field in the "Name" tab no longer creates a 500 error when it is blank and you click save.

3.  Blank Jobs and Schools are now allowed.  The user can remove all jobs or schools, and does not prevent her from deleting her last job.

4.  There is now better client side validation:  Now the "Institution" field is required ONLY if "Department" or "Job Title" are filled in.

4.  The "Remove" link now has a modal window asking if you want to delete.  If you click yes, the job or school is deleted and the profile is saved.

6.  Any blank Institutions are ignored on save.

Side Effect
--------------

1.  When a user removes a job or school, it will delete that job unless it is the last job.  If it is the last remaining job or school, all the fields for that job or school will be cleared out, and a blank school or job list is saved to the database.

2.  Remove also saves the entire object.  So, if you Create "Job 1" and "Job 2", modify "Job 1", then click "remove" on "Job 2" and confirm, The change to "Job 1" is also saved.

3.  If in the previous case, "Job 1" is invalid (no "Institution" field, bad start year, etc), then clicking "remove" on "Job 2" will not be saved. Instead the user will be alerted that "Institution is Required" and She will have to make "Job 1" valid and then click "Save" to save both changes.  This is not ideal behavior, but since saving each object individually is beyond the scope of this PR, and overall, this is a usability improvement, it is considered currently acceptable.

billyhunt commented on framework/auth/core.py:
```diff
@@ -56,7 +56,6 @@ def string_required(value):
 
 
 def validate_history_item(item):
-    string_required(item.get('institution'))
```
Removed because validation is more complicated and occurs on client side.  (institution is only required when there is a list item, otherwise it is not.

sloria replied:
Does it have to be this way? By the time the data reach the server, the data should be valid, correct? In other words, `institution` should always be present in the payload sent to the server.

billyhunt replied:
When you delete the entire object, then the institution will be cleared along with all other fields.

billyhunt commented on tests/test_models.py:
```diff
@@ -130,9 +130,9 @@ def test_validate_jobs_valid(self):
+        self.user.jobs = []
+        self.user.save()
+        assert_equal(self.user.jobs, [])
 
```
See previous comment.

billyhunt commented on website/static/js/profile.js:
```diff
@@ -436,6 +446,32 @@ var NameViewModel = function(urls, modes, preventUnsaved, fetchCallback) {
+    }
+
+};
+
```
Overriding BaseViewModel submit button to check for Full Name in profile.

billyhunt commented on website/templates/include/profile/jobs.mako:
```diff
@@ -96,7 +95,7 @@
 
             <div>
-                <a class="btn btn-default" data-bind="click: addContent">
+                <a class="btn btn-default" data-bind="disable, click: addContent">
```
disable is a knockout binding that disables a button if the validation requirements are not met.  In this case, if the job or school is empty.

billyhunt commented on website/static/js/profile.js:
```diff
@@ -577,9 +613,45 @@ var ListViewModel = function(ContentModel, urls, modes) {
         return true;
     });
+
+    self.institutionsEmpty = ko.computed(function() {
```
@sloria, you had some questions about "institution" being in the ListView.  "institution" is the generic name for the title field of both schools and jobs in our profile.  It doesn't seem like we need to refactor unless we are changing the structure of the objects.  Let me know if you think differently.

sloria commented on website/static/js/profile.js:
```diff
@@ -587,9 +636,36 @@ var ListViewModel = function(ContentModel, urls, modes) {
+        return false;
+    });
+
+    self.jobsOrSchools = ko.computed(function() {
```
This data does not belong in the base class (`ListViewModel`)--these are the responsibility of the concrete class.

sloria commented on website/static/js/profile.js:
```diff
@@ -587,9 +636,36 @@ var ListViewModel = function(ContentModel, urls, modes) {
+
+    self.missingRequiredfield = ko.computed(function() {
+        for (var i=0; i<self.contents().length; i++) {
+            if (self.jobsOrSchools() === JOBS) {
```
See previous comment re: encapsulation--the base `ListViewModel` should not have to know about `JOBS` and `SCHOOLS`.

sloria commented:
> The Add Institution button will be disabled if there is a blank object. It becomes enabled as soon as any field is filled in.

We decided [a while back](https://github.com/CenterForOpenScience/osf.io/issues/1035) that this is an anti-pattern because it doesn't make clear to the user what is wrong with their input. To follow the current pattern, the user should be allowed to click "Add Institution" and receive feedback about what was wrong with their input, if appropriate.

sloria commented on website/static/js/profile.js:
```diff
@@ -446,6 +456,32 @@ var NameViewModel = function(urls, modes, preventUnsaved, fetchCallback) {
+NameViewModel.prototype.submit = function() {
+    if (this.hasValidProperty() && this.isValid()) {
+        // Name on Name view is empty
+        if (this.nameFieldEmpty()) {
```
This shouldn't be necessary--can you use the HTML `required` attribute on the name field instead?

sloria commented on website/static/js/profile.js:
```diff
@@ -675,24 +763,55 @@ ListViewModel.prototype.unserialize = function(data) {
     return {contents: contents};
 };
 
+ListViewModel.prototype.submit = function() {
```
This override should not be necessary; `self.isValid()` should take care of any validation before submitting the form.

billyhunt replied:
If no fields are filled in, then you don't want to flag it because it should blank the object, hence the added complexity.

sloria commented:
Pass finished. The main things to improve about this approach:

- Only valid data must be allowed to reach the server.  This means that the same validation should happen on the server and client side. Partial data (e.g. an "Education" item without the institution specified should not even be sent to the server).
- Simplify how validation happens. Notice that you have overridden `submit` in two subclasses; this is a code "smell"--a hint of bad design (see the [open/closed principle](http://www.oodesign.com/open-close-principle.html)). See if you can use built-in browser validation (HTML `required`) or KO-validation so that `isValid()` returns `false` when the data are invalid rather than checking the values in `submit`.

sloria commented:
It seems that the "Institution" field has been made required since this PR has been sent, so I haven't been able to reproduce https://github.com/CenterForOpenScience/osf.io/issues/2412 .

billyhunt commented:
I updated the behavior of the add button so it will give user input on how to move forward.  

I also simplified the validation, using knockouts internal validation for institutions using required: onlyIf

Finally, blank objects are not saved on the server.

sloria commented on website/static/js/profile.js:
```diff
@@ -708,6 +763,18 @@ var JobViewModel = function() {
 
     var validated = ko.validatedObservable(self);
+
+    self.institutionObjectEmpty = ko.computed(function() {
```
This can be a `pureComputed`

sloria commented on website/static/js/profile.js:
```diff
@@ -708,6 +763,18 @@ var JobViewModel = function() {
     var validated = ko.validatedObservable(self);
+
+    self.institutionObjectEmpty = ko.computed(function() {
+        if (self.institution() === '' && self.department() === '' && self.title() === '') {
```
This can be simplified to:

```js
return self.institution() && self.department() && self.title();
```

sloria commented on website/static/js/profile.js:
```diff
@@ -693,10 +737,21 @@ var JobViewModel = function() {
+        trimmed: true,
+        required: {
+            onlyIf: function() {
+                if (!!self.department() || !!self.title()) {
```
This can be simplified to

```js
return !!self.department() || !!self.title();
```

Remember, these are antipatterns:

```js
if (condition) { return true; }
return condition ? true : false;

// both can be simplified to just
return condition
```

sloria commented on website/static/js/profile.js:
```diff
@@ -647,12 +664,33 @@ var ListViewModel = function(ContentModel, urls, modes) {
+    }
+    else {
+        this.changeMessage(
+            'Institution Field is Required',
```
Sentence case is preferred: https://centerforopenscience.github.io/osf-style/language.html#capitalization

sloria commented on website/static/js/profile.js:
```diff
@@ -591,6 +603,11 @@ var ListViewModel = function(ContentModel, urls, modes) {
         return self.contents().length > 1;
     });
+
+    self.contentsLength = ko.computed(function() {
```
This isn't necessary; just check `self.contents().length;` 

Avoid unnecessary computeds; they add extra...computation..when observables are updated.

sloria commented on website/static/js/profile.js:
```diff
@@ -676,15 +714,21 @@ ListViewModel.prototype.unserialize = function(data) {
 ListViewModel.prototype.serialize = function() {
     var contents = [];
-    if (this.contents().length !== 0 && typeof(this.contents()[0].serialize() !== undefined)) {
+    if (this.contents().length !== 0 && typeof(this.contents()[0].serialize()) !== 'undefined') {
```
Check against undefined like so:

```js
// null or undefined
this.contents()[0].serialize() != null

// undefined only
this.contents()[0].serialize() === undefined
```

https://github.com/rwaldron/idiomatic.js/#type

sloria commented on website/static/js/profile.js:
```diff
@@ -647,12 +664,33 @@ var ListViewModel = function(ContentModel, urls, modes) {
-    this.contents.splice(idx, 1);
+    //  If there is more then one model, then delete it.  If there is only one, then delete it and add another.
+    if (this.contentsLength() > 1) {
+        this.contents.splice(idx, 1);
```
This happens in both the `if` and `else` branch, so it can be pulled out of here; then you'll only need a single `if` branch (no need for `else`).

sloria commented:
Pass finished.

sloria commented on tests/test_models.py:
```diff
@@ -134,6 +134,9 @@ def test_validate_jobs_institution_empty(self):
             self.user.save()
+        self.user.jobs = []
+        self.user.save()
+        assert_equal(self.user.jobs, [])
```
This doesn't seem to be testing anything. Remove?

sloria commented on website/static/js/profile.js:
```diff
@@ -351,7 +349,12 @@ var NameViewModel = function(urls, modes, preventUnsaved, fetchCallback) {
+    self.full = koHelpers.sanitizedObservable().extend({
+        trimmed: true,
+        required: true,
+        message: 'Full Name Required'
```
Is this message displayed anywhere? When I submit an empty full name, I don't see this message.

sloria commented on website/static/js/profile.js:
```diff
@@ -580,6 +583,15 @@ var ListViewModel = function(ContentModel, urls, modes) {
         return self.contents().length > 1;
     });
 
+    self.institutionObjectsEmpty = ko.computed(function(){
```
This can be a `pureComputed`.

sloria commented on website/static/js/profile.js:
```diff
@@ -734,6 +795,12 @@ var SchoolViewModel = function() {
     var validated = ko.validatedObservable(self);
+
+    //In addition to normal knockout field checks, check to see if institution is not filled out when other fields are
+    self.institutionObjectEmpty = ko.purecomputed(function() {
```
Typo here is breaking the profile page: `pureComputed`.

sloria commented:
I still see a 500 error when submitting an empty form in Employement or Education.

sloria commented:
Please do another functional review and address the above bugs. Once that is addressed, I will do another pass.

billyhunt commented on website/templates/include/profile/jobs.mako:
```diff
@@ -17,20 +17,19 @@
                         <span>Position {{ $index() + 1 }}</span>
-                        <span data-bind="visible: $parent.hasMultiple()">
+                        <span data-bind="visible: $parent.contentsLength() > 1">
                             [ drag to reorder ]
```
That 500 error was happening when I removed the observable contentsLength (which you asked to be removed).   The .length property was not available for the knockout variable at the right time.  hasMultiple was redundant however, so I deleted it and am using contentsLength instead, so we are still -1 knockout variable.  :facepunch:

lyndsysimon replied:
:facepunch:

lyndsysimon commented on website/static/js/profile.js:
```diff
@@ -580,6 +582,15 @@ var ListViewModel = function(ContentModel, urls, modes) {
     });
 
+    self.institutionObjectsEmpty = ko.pureComputed(function(){
+        for (var i=0; i<self.contents().length; i++) {
```
As I understand this loop, the function will return `true` if the first item in the list is empty. I believe the `false` return should be inside the conditional, not outside.

billyhunt replied:
This loop checks if there are any blank institution objects within the list to prevent users from continuing to add blank institutions to the view (madness!).  So if it finds any blank institutions, it exits the loop with a True, if not, its False City!

lyndsysimon commented on website/static/js/profile.js:
```diff
@@ -693,10 +731,19 @@ var JobViewModel = function() {
+        trimmed: true,
+        required: {
+            onlyIf: function() {
+               return !!self.department() || !!self.title();
```
@sloria - are "bang-bangs" in keeping with our style guide? I don't see it mentioned.

sloria replied:
They do not conflict with our style guide. While I think they're unnecessary, I don't have a strong preference against them.

billyhunt replied:
For some reason, a :exclamation: does not work with knockout require, so :bangbang: are back in.

sloria replied:
The reason is that `!` is for negation, so your logic would be wrong. `!!` just makes truthy values evaluate to `true`.

```js
var truthyString = 'lol';
console.log(!!truthyString);  // true
```

In most cases, the `!!` doesn't make a difference.

lyndsysimon commented:
CR complete - one potential bug identified.

lyndsysimon commented:
What I thought might have been a bug was not a bug. Good to go :)

:star:

sloria commented on website/static/js/profile.js:
```diff
@@ -647,12 +660,48 @@ var ListViewModel = function(ContentModel, urls, modes) {
-    this.contents.splice(idx, 1);
+    var self = this;
+
+        bootbox.confirm({
```
Unnecessary indentation.

sloria commented on website/static/js/profile.js:
```diff
@@ -647,12 +660,48 @@ var ListViewModel = function(ContentModel, urls, modes) {
+
+        bootbox.confirm({
+            title: 'Remove Institution?',
+            message: 'Are you sure you want to remove this Institution?',
```
"Institution" should not be capitalized.

sloria commented:
I'll make the final minor changes when I merge.
